0005-ordered-quantification.md

```diff
@@ -0,0 +1,626 @@
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in the set above, sorted by the natural ordering of `int` values,
+is simple using a sequence comprehension expression: `seq i: int | i in s`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Both of the proposed new features depend on another new and more fundamental concept: quantification ordering. 
+Therefore, allow me to outline this concept first before then describing `foreach` loops and sequence comprehensions.
+
+## Quantification ordering
+
+Several existing Dafny features support a common syntax for quantifying over one or more variables, internally referred to
+as a "quantifier domain". For example, the `x: nat, y: nat | x < 2 && y < 2` portion of the set comprehension 
+`set x: nat, y: nat | x < 2 && y < 2 :: (x, y)`. The boolean expression after the `|` is referred to as the *range*,
+and I will label the `x: nat` and `y: nat` sections as *quantifier variable declarations*.
+We can augment this existing syntax with a notion of ordering, and allow quantified variables to bind to duplicate values. The key points are:
+
+* Any quantifier domain defines a potentially-infinite, partially-ordered set of *quantifier bindings*.
+* The quantified variable declarations define the values each binding maps to each variable, AND how these bindings are ordered.
+* The range expression after the `|` restricts the quantification to a subset of these bindings, but does not influence their ordering.
+
+A quantifier domain only guarantees an ordering of bindings, 
+but is NOT prescriptive on how to enumerate this domain at runtime, if it is compiled!
+This is consistent with existing expressions such as `set x: real | x in myFiniteSet && P(x)`: 
+although the unfiltered domain of `real` values is not enumerable, the bound `x in myFiniteSet` is, 
+and at runtime this comprehension is calculated by enumerating the contents of `myFiniteSet` 
+and filtering out values that do not satisfy `P(x)`. 
+The new features that are affected by quantification ordering will behave similarly: 
+`seq x: real | x in mySet && P(x)` is calculated via the same filtered enumeration, 
+but collected into a sorted sequence instead.
+
+There will be three supported variable declaration styles, and others could potentially be added in the future:
+
+1. `x [: T]`
+
+    In the existing syntax for quantifier variables (where the type `T` may be explicitly provided or omitted and inferred),
+    the bindings will be ordered according to the *natural ordering* of the type `T`. Not all Dafny types will
+    have such a natural ordering defined, but at a minimum `int` and `real`, and any subset type or newtype based on them,
+    will use the natural mathematical ordering.
+    `x: int | x in {2, 5, -4}`, for example, would bind `x` to `-4`, `2`, and `5` in that order.
```

Comment:
In which case is the natural ordering *needed*?  I'm surprised that we still need an ordering at all: for the example above we can write `seq x: real <- sorted(mySet) | P(x)`, right?  And for this one the libraries can provide a conversion function from sets?

Reply:
Also, how will this work with types that don't have comparisons?  I'm assuming that types that don't support `==` also don't support comparisons, and yet this feature will let me compare arbitrary values: `var aLeB := (seq x | x in {a, b}) == [a, b]`

Reply:
> Also, how will this work with types that don't have comparisons? I'm assuming that types that don't support `==` also don't support comparisons, and yet this feature will let me compare arbitrary values: `var aLeB := (seq x | x in {a, b}) == [a, b]`

This is a very good point - I've been assuming in all my examples that the types were equality-supporting to focus on defining the semantics. That means in some cases you can't even create a compiled `{a, b}` value in the first place. I'll call that out more explicitly and watch out for under-specified examples.

But yes, not all types will have natural orderings (any reference type for example) so some types will require `<-`.